[Search] Improve generic search performance

VWoeltjen opened the pull request:
Addresses #141.

Summary of changes:
* Broadcast mutations on one global topic to avoid the need to attach mutation listeners per-object; this allows components which are interested in observing all mutations (such as GenericSearch) to maintain a single listener, instead of one per object, reducing memory usage.
* Cherry-pick a fix to `throttle` from #104 to resolve a race condition there.
* Change approach to indexing in GenericSearch to (a) avoid the use of capabilities, and (b) spread out requests for objects over time. In both cases, the intent is to reduce the CPU utilization and/or time spent associated with initial indexing. More detail is in comments on #141
* Add an indicator to `example/profiling` to report number of digest cycles per second, to aid in diagnosing performance related issues such as this.

### Author Checklist

1. Changes address original issue? Y
2. Unit tests included and/or updated with changes? Y
3. Command line build passes? Y
4. Expect to pass code review? Y

larkin commented on platform/core/src/capabilities/MutationCapability.js:
```diff
@@ -115,11 +120,19 @@ define(
 
+            function notifyListeners(model) {
+                // Broadcast a general event...
+                generalTopic.notify(domainObject);
```
These function names speak for themselves and don't need comments.

larkin commented on platform/core/src/services/Throttle.js:
```diff
@@ -56,12 +61,14 @@ define(
              */
             return function (fn, delay, apply) {
-                var activeTimeout;
+                var promise, // Promise for the result of throttled function
```
Lots of unnecessary comments here-- duplicating what's already stated by the code.

larkin commented on platform/search/src/services/GenericSearchProvider.js:
```diff
@@ -110,83 +125,50 @@ define(
+                        ' seconds.'
+                    ].join(''));
+                } else {
+                    Object.keys(toRequest)
```
Seems like toRequest should not be an object: you're constantly coercing it to an array (spendy) and never using it to check for membership. it would perform better as an array.

larkin commented:
@VWoeltjen left notes, in general there is a lot of comments that can be removed-- the code already says "how" it works and the comments just get in the way.

The queue implementation is really not the right data structure (should just be an array).  

Assigning back to you for fixes.

VWoeltjen commented:
Pushed changes to open141:

* Reduced flush interval to 0 to reflect latest from discussion on #141
* Changed data structure for ids pending index to array; added second set-like object to keep values in this array unique.
  * The `indexed` set _almost_ does this already, except that it needs to be cleared to allow reindexing of mutated objects. The two sets do reflect different things: `indexed` is objects that have ever been indexed, and `pendingIndex` is specifically the set of items that are in the queue to index.
* Removed some comments

Reassigning to @larkin for rereview

larkin commented:
# Review Checklist

1. Changes appear to address issue? Y
2. Appropriate unit tests included? Y
3. Code style and in-line documentation are appropriate? Y
4. Commit messages meet standards? Y
5. Project-specific information isolated to appropriate branches? Y

# Integration Checklist

1. Automated build passing after merge? Y
2. No merge conflicts (or resolution trivial/obvious)? Resolved one trivial conflict.
3. All master branches up-to-date after merge? Y
4. Project-specific information isolated to appropriate branches? Y
